### ISSUE TITLE
extract RouteMiddleware from Application

### ISSUE THREAD

**Issue by renanbr** [zendframework/zend-expressive#216]
Still regarding routing. Why not to move `Application::routeMiddleware()` and related methods to some real middleware? `Zend\Expressive\Application` could be kept as simple as a middleware conductor, less responsibility, less code...

thanks zend for this nice project!

**Comment by weierophinney** [comment]
We debated this heavily before settling on the current solution.

An `Application` instance *without* the routing middleware is essentially Stratigility's `MiddlewarePipe`. If you want that, use that.

The reasons:

- The container then needs to be injected in the middleware *and* the application, which complicates initial setup. Having multiple container-aware objects is a code-smell.
- The routing middleware typically needs to be added manually to the Application (to allow middleware pipelines); now users are forced to juggle multiple objects.
- Where is routed middleware added? The router? the route middleware? the application? If there's more than one answer, there's then more than one way to accomplish the same thing, which leads to developer confusion.
- In separating the router interface and implementations out, we already ran into some issues of consistency: how do we prevent duplicate routes? how do we ensure route result observers are triggered?

Mandating that the `Application` composes the routing middleware:

- Ensures that routing is built-in, without requiring developer action. Without it, as noted earlier, it's just a `MiddlewarePipe`. *Routing is a key value-add of Expressive.*
- Ensures consistency for selected application responsibilities (route duplication detection; route result observer notification; middleware marshaling; etc.)
- Ensures a single point-of-entry for setting up the application middleware.

This boils down to two words: developer experience.

Consider the following, which is a basic use case:

```php
use Zend\Diactoros\Uri;
use Zend\Expressive\AppFactory;
use Zend\Expressive\Template\TemplateRendererInterface;

$app = AppFactory::create();
$app->get(...);
```

If we separate the routing middleware to another object, we have several things to consider:

- Where are routed middleware added: the routing middleware? the application?  the router?
- How and when is the routing middleware added to the application?

Let's look at some ideas:

```php
use Some\Container;
use Zend\Expressive\Application;
use Zend\Expressive\Router\RouterInterface;
use Zend\Expressive\Router\RoutingMiddleware;
use Zend\Expressive\Template\TemplateRendererInterface;

$container = new Container;
$app = $container->get(Application::class);
$router = $container->get(RouterInterface::class);
$routing = $container->get(RoutingMiddleware::class);

// Do we add routed middleware to the Application?
// If so, that means that the Application is proxying to the routing middleware
// and/or router. Or maybe to both?
$app->get(...);

// Or do we add routed middleware to the Router?
// This would make the RouterInterface far bigger and more complicated, and
// potentially lead to differences between implementations. Also: is the router
// already composed in the Application instance, or would we need to add it
// manually?
$router->get(...);

// Or do we add routed middleware to the RoutingMiddleware?
// This would require the RoutingMiddleware be container-aware, and also
// require manually adding it to the application to ensure we craft the
// pipeline we want.
$routing->get(...);
$app->pipe($routing);
```

Add to this some other issues:

- How is middleware specified? Are service names okay? what about callables? What about arrays of middleware? Are route and pipeline middleware specified in the same ways, or are there differences?
- what object is responsible for triggering route result observers?
- what object is responsible for error handling and marshaling error handlers?  (this is why the `Application` must *also* be container aware in all scenarios.)
- what object is responsible for marshaling the initial request and response?  (this one likely can only ever be answered by the `Application` class, but it's a fair question)
- when using the component standalone, how much domain knowledge is needed in order to start a usable application? While a clean separation of responsibilities is always a good goal, too many initial layers leads to more requisite knowledge in order to begin using the component.

Additionally, separating the routing middleware introduces additional support vectors:

- Standalone users unaware that they need to add the middleware to the application instance will be a constant source of issues.
- Alternate routing middleware implementations will introduce inconsistency in operation, leading to increased support requests.

Essentially, pushing the methods for attaching routes to a lower layer complicates usage and introduces potential inconsistency (e.g., route de-duplication, route result observer notification, middleware marshaling).  Many of these issues could be handled in the `ApplicationFactory`, which orchestrates how the layers work together. However, while we are encouraging usage of a DI container with Expressive, using one *does not require that you use the Expressive skeleton, nor the `ApplicationFactory`*, which means that the *manual* usage of Expressive needs to be considered equally. 

The solution to this is to create a [poka-yoke](https://en.wikipedia.org/wiki/Poka-yoke), or a design that encourages a single workflow.  As such, we chose to implement the `Application` class as a [facade](https://en.wikipedia.org/wiki/Facade_pattern) (as in the *design pattern*), and have it as a single entry point to the various interactions.

tl;dr: Separating the routing middleware to another object or layer complicates usage significantly for little gain other than technical purity, and introduces many new support vectors that will hinder progress on the project.

**Comment by renanbr** [comment]
"_Verba volant, scripta manent_"
Thanks for your time and this detailed explanation!
